Vault cert provider

optiz0r opened the pull request:
Add support for deploying arbitrary certificates issued by vault for consumption by other applications managed by puppet
* Includes `vault_cert` resource type and provider
* Includes a full test suite
* Includes a manifest to deploy prerequisites and handle optional orphaned resource purging
* Updates REFERENCE and README docs with usage example

Fixes #2

southalc commented:
Sorry it's taken me a bit to get back to this.  I like the provider/type implementation and appreciate the included test suite.  Running "pdk test unit" completes for me with no failures.

I checked the module with "pdk validate" and there are a number of suggestions to align with coding standards.  I generally try to make all my module code pass validation cleanly.  I also wondered about the choice of file name of "ruby.rb"?  The recommended practice seems to be to name the type/provider files after the type.
https://puppet.com/docs/puppet/7/create_types_and_providers_resource_api.html#create_types_and_providers_resource_api

Would you be able to address these?  I appreciate your work on this feature.

optiz0r commented:
Certainly. I'll see what checks are falling and make them pass. 

I've done a few new host builds over the weekend and spotted some ordering issues. My vault ca cert is added to the system trust store during a puppet run, and if vault_cert provider tries to run before that happens the certs won't be placed properly. It does resolve itself after another one/two puppet runs, but I will see if there's anything I can do to make this more graceful. I think I can also add some autorequires to help simplify the puppet manifests.

I was completely unaware there is a new Resources Api method to implement types and providers. I will take a look and see if this is worth rewriting to use the new approach. An initial read of the docs makes it look like the read only attributes are now a built in feature, which will save some hacks. The docs do not show an example of delayed flushing though, so I'll need to do some digging to see if this is possible under the new api.

optiz0r commented:
Validate and test suite now run clean. I've fixed a couple of issues
- info file was being written with wrong mode on initial puppet run and fixed on subsequent runs
- chain/crt/key files were not being placed until second run due to a crash
- using `puppet resource vault_cert test ensure=absent` now correctly shows the final state rather than previous state

Still to look at:
- autorequires for parent directories where info/chain/crt/key files will be placed (definitely before merging)
- renaming the provider from :ruby to something else? (optional before merging)
- refactoring to using newer resources api? (optional before merging)
